### PR TITLE
Sampling op

### PR DESCRIPTION
**Pull request by wangyems**
### Description

Sampling op for cpu and cuda
support huggingface case and custom case

### CODE REVIEW

**Comment by yufenglee** [comment]
>       }

if it is +inf, set the max to float max

---
Refers to: onnxruntime/core/providers/cpu/generator/random.cc:245 in 6ff8f23. [](commit_id = 6ff8f23552f9ed5b841e04e295fbc303da54221f, deletion_comment = False)

File: onnxruntime/contrib_ops/cpu/transformers/sampling.cc [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  ORT_ENFORCE(info.GetAttr<ONNX_NAMESPACE::GraphProto>("decoder", &proto).IsOK());
+  ORT_IGNORE_RETURN_VALUE(proto);
+}
+
+Status Sampling::SetupSubgraphExecutionInfo(const SessionState& session_state,
+                                            const std::string& attribute_name,
+                                            const SessionState& subgraph_session_state) {
+  const auto& node = Node();
+  if (parameters_.model_type == IGenerationParameters::kModelTypeGpt) {  // GPT-2
+    if (attribute_name == "decoder") {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/sampling.cc:58]: >decoder [](http://example.com/codeflow?start=27&length=7)

This PR [https://github.com/microsoft/onnxruntime/pull/13943](https://github.com/microsoft/onnxruntime/pull/13943) handles init_decoder and decoder separately to improve the performance of logits computation of init_decoder. This sampling can leverage the same change.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -1114,6 +1121,42 @@ ONNX_MS_OPERATOR_SET_SCHEMA(GreedySearch, 1,
...
+                                .Input(4, "vocab_mask", "Mask of vocabulary. Words that masked with 0 are not allowed to be generated, and 1 is allowed. Shape is (vacab_size)", "I", OpSchema::Optional)
+                                .Input(5, "prefix_vocab_mask", "Mask of vocabulary for first step. Words that masked with 0 are not allowed to be generated, and 1 is allowed. Shape is (batch_size, vocab_size)", "I", OpSchema::Optional)
+                                .Input(6, "attention_mask", "Custom attention mask. Shape is (batch_size, sequence_length)", "I", OpSchema::Optional)
+                                .Input(7, "presence_mask", "presence penalty mask. Shape is (batch_size, vocab_size)", "I", OpSchema::Optional)
+                                .Output(0, "sequences", "Word IDs of generated sequences. Shape is (batch_size, max_sequence_length)", "I")
+                                .Output(1, "filtered_logits", "filtered logits as input to the mutinomial function . Shape is (batch_size, vocab_size)", "T", OpSchema::Optional)
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1153]: >.Output(1, "filtered_logits", "filtered logits as input to the mutinomial function . Shape is (batch_size, vocab_size)", "T", OpSchema::Optional) [](http://example.com/codeflow?start=32&length=145)

this is for debug?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -1114,6 +1121,42 @@ ONNX_MS_OPERATOR_SET_SCHEMA(GreedySearch, 1,
...
+                                .Attr("temperature", "temperature for sampling", AttributeProto::FLOAT, 1.0f)
+                                .Attr("top_p", "top_p for sampling", AttributeProto::FLOAT, 0.0f)
+                                .Attr("filter_value", "filter value for top_p", AttributeProto::FLOAT, -1e20f)
+                                .Attr("min_tokens_to_keep", "min_tokens_to_keep", AttributeProto::INT, static_cast<int64_t>(0))
+                                .Attr("presence_penalty", "presence penalty for sampling", AttributeProto::FLOAT, 0.0f)
+                                .Attr("custom", "absence penalty for sampling", AttributeProto::INT, static_cast<int64_t>(0))
+                                .Attr("model_type", "model type: 0 for decoder only like GPT-2; 1 for encoder decoder like Bart", AttributeProto::INT, static_cast<int64_t>(0))
+                                .Attr("encoder", "The subgraph for initialization of encoder and decoder. It will be called once before decoder subgraph.", AttributeProto::GRAPH, OPTIONAL_VALUE)
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1138]: need to add an attribute 'vocab_size' to leverage the padding optimization for vocab:
[https://github.com/microsoft/onnxruntime/pull/13842](https://github.com/microsoft/onnxruntime/pull/13842)
[https://github.com/microsoft/onnxruntime/pull/13686](https://github.com/microsoft/onnxruntime/pull/13686)

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 2 above)
Comment (wangyems) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1153]: yes

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+  int batch_id = index / vocab_size;
+  int start_index = batch_id * vocab_size;
+
+  int count = vocab_idx;
+  float sum = 0.0f;
+  while (count >= 0) {
+    sum += d_sorted_logits_in[start_index];
+    ++start_index;
+    --count;
+  }
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:531]: it looks like quite slow. We need following PRs to optimize it. One thing i can think of now is to use cub's scan.

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 5]
```diff
@@ -0,0 +1,775 @@
...
+  int warp_size = 32; //at::cuda::warp_size();
+  int requiredWarps = (vocab_size + warp_size - 1) / warp_size;
+  int requiredThreads = std::min(maxThreads, requiredWarps * warp_size);
+  int requiredShared = requiredThreads * sizeof(float);
+
+  // bugbug: randomize d_sampled
+  dim3 block(requiredThreads);
+  dim3 grid(std::min(batch_size, numSM * 4));
+
+  if (block.x == 1024) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:755]: > if (block.x == 1024) { [](http://example.com/codeflow?start=1&length=23)

what if it is not 1024?

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 6]
```diff
@@ -0,0 +1,775 @@
...
+    scalar_t val;
+    for (int cat = threadIdx.x; cat < categories; cat += blockDim.x) {
+      val = dist[curDist * stride_dist + cat * stride_categories];
+      // CUDA_KERNEL_ASSERT(!at::_isnan(val));
+      // CUDA_KERNEL_ASSERT(!_isinf(val));
+      // CUDA_KERNEL_ASSERT(!(val < zero));
+      sum = sum + static_cast<accscalar_t>(val);
+    }
+    // threadIdx.x == 0 has the sum value from this
+    sum = BlockReduce(tmp_storage).Reduce(sum, cub::Sum()); // sum = cuda_utils::BlockReduceSum(sum, smem);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:632]: sum should be 1.f as the input is SoftMax-ed

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 7]
```diff
@@ -0,0 +1,775 @@
...
+  int numSM = props.multiProcessorCount;
+  int maxThreads = props.maxThreadsPerBlock;
+  int warp_size = 32; //at::cuda::warp_size();
+  int requiredWarps = (vocab_size + warp_size - 1) / warp_size;
+  int requiredThreads = std::min(maxThreads, requiredWarps * warp_size);
+  int requiredShared = requiredThreads * sizeof(float);
+
+  // bugbug: randomize d_sampled
+  dim3 block(requiredThreads);
+  dim3 grid(std::min(batch_size, numSM * 4));
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:753]: >batch_size, numSM * 4 [](http://example.com/codeflow?start=21&length=21)

why not juse use batch_size?

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 8]
```diff
@@ -0,0 +1,775 @@
...
+
+// Ref: https://github.com/pytorch/pytorch/blob/release/1.13/aten/src/ATen/native/cuda/MultinomialKernel.cu
+template <typename scalar_t, typename accscalar_t, unsigned TPB>
+__global__ void sampleMultinomialOnce(int64_t* dest,
+                                      int distributions,
+                                      int categories,
+                                      scalar_t* sampled,
+                                      scalar_t* dist,
+                                      int stride_dist, // dist->stride(0)
+                                      int stride_categories, // dist->stride(1)
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:606]: >stride_categories [](http://example.com/codeflow?start=42&length=17)

I don't understand why stride_categories is batch_size instead of 1?

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 5 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:755]: we just need to add that template when needed. based on most common vocab value 1024 is enough

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 4 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:531]: will add a TODO

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 6 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:632]: good catch! we can safely remove this step since this function is only used for topp

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 7 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:753]: this function is basically copied from pytorch cuda implementation so I guess there might be some advantages here

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 5 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:755]: the template for AllReduce is removed

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 8 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:606]: good catch! it should be one

**Review by yufenglee — APPROVED** [review]
